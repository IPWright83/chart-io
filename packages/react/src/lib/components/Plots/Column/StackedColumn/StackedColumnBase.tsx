import * as d3 from "@d3-chart/d3";
import type { IColor, IDatum, IEventPlotProps } from "@d3-chart/types";
import { useEffect, useState } from "react";
import { useSelector, useStore } from "react-redux";
import type { Transition } from "@d3-chart/d3";

import { chartSelectors, eventActions, IState } from "../../../../store";
import { ensureBandScale, ensureNoScaleOverflow, ensureValuesAreUnique } from "../../../../utils";
import { useRender } from "../../../../hooks";

import { getDropline } from "../getDropline";
import { getParentKey } from "./getParentKey";
import { renderCanvas } from "../../renderCanvas";
import { useTooltip } from "../useTooltip";

export interface IStackedColumnBaseProps extends Omit<IEventPlotProps, "y"> {
    /**
     * This is an internally used function to allow the scatter plot to render to a virtual canvas
     */
    renderVirtualCanvas?: (update: Transition<Element, unknown, any, unknown>) => void;
    /**
     * The set of y fields to use to access the data for each plot
     */
    ys: Array<string>;

    /**
     * The set of colors to use for the different plot
     */
    colors?: Array<IColor>;
}

/**
 * Represents a Column Plot
 * @param  props       The set of React properties
 * @return             The Column plot component
 */
export function StackedColumnBase({
    x,
    ys,
    colors,
    interactive = true,
    onMouseOver,
    onMouseOut,
    onClick,
    layer,
    canvas,
    renderVirtualCanvas,
}: IStackedColumnBaseProps) {
    const [focused, setFocused] = useState(null);
    const store = useStore();

    const data = useSelector((s: IState) => chartSelectors.data(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const xScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, x));
    const yScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, ys[0]));
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const animationDuration = useSelector((s: IState) => chartSelectors.animationDuration(s));

    const strokeColor = theme.background;
    const setTooltip = useTooltip(store.dispatch, x);

    useEffect(() => {
        if (!focused) return;

        const selection = d3.select(focused.element).style("opacity", theme.series.selectedOpacity);
        const dropline = getDropline(selection, xScale);
        store.dispatch(eventActions.addDropline(dropline));

        // Clean up operations on exit
        return () => {
            selection.style("opacity", theme.series.opacity);
            store.dispatch(eventActions.removeDropline(dropline));
        };
    }, [store.dispatch, focused, xScale, theme.series.opacity, theme.series.selectedOpacity]);

    // prettier-ignore
    useRender(() => {
        if (ensureBandScale(xScale, "StackedColumnBase") === false) return null;
        ensureValuesAreUnique(data, x, "StackedColumnBase");
        ensureNoScaleOverflow(yScale, data, ys, "StackedColumnBase");

        // Create the stacked variant of the data
        const keys = ys;
        // @ts-ignore: TODO: Fix this
        const stackedData = d3.stack().keys(keys)(data);
        const colorScale = d3.scaleOrdinal().domain(keys).range(colors);

        const groupJoin = d3.select(layer.current)
            .selectAll<SVGGElement, IDatum>("g")
            .data(stackedData);

        // Clean up old stacks
        groupJoin.exit().remove();

        const join = groupJoin
            .enter()
            .append("g")
            .merge(groupJoin)
            .selectAll(".column")
            .data((d) => d);

        join.exit().remove();
        const enter = join
            .enter()
            .append("rect")
            .attr("class", "column")
            .attr("x", (d) => xScale(d.data[x]))
            .attr("y", () => yScale.range()[0])
            .attr("height", 0)
            // @ts-expect-error: scale.bandwidth() has already been protected against using ensureBandScale()
            .attr("width", xScale.bandwidth())
            .style("fill", (d, i, elements) => {
                const key = getParentKey(elements[i]);
                return colorScale(key)?.toString();
            })
            .style("stroke", strokeColor?.toString())
            .style("opacity", theme.series.opacity);

        const update = join
            .merge(enter)
            .on("mouseover", function (event: MouseEvent, d) {
                // istanbul ignore next
                if (!interactive) return;

                onMouseOver && onMouseOver(d.data, this as Element, event);
                setFocused({ element: this, event, datum: d.data });
                setTooltip({
                    datum: d.data,
                    event,
                    fillColors: ys.map((y) => colorScale(y) as IColor),
                    ys,
                });
            })
            .on("mouseout", function (event: MouseEvent, d) {
                // istanbul ignore next
                if (!interactive) return;

                onMouseOut && onMouseOut(d.data, this as Element, event);
                setFocused(null);
                setTooltip(null);
            })
            .on("click", function (event: MouseEvent, d) {
                // istanbul ignore next
                if (!interactive) return;

                onClick && onClick(d.data, this as Element, event);
            })
            .transition("position")
            .duration(animationDuration / 2)
            .attr("x", (d) => xScale(d.data[x]))
            // @ts-expect-error: scale.bandwidth() has already been protected against using ensureBandScale()
            .attr("width", xScale.bandwidth())
            .style("fill", (d, i, elements) => {
                const key = getParentKey(elements[i]);
                return colorScale(key)?.toString();
            })
            // @ts-expect-error: Looks like the type defs are wrong missing named transitions
            .transition("height")
            .duration(animationDuration / 2)
            .delay(animationDuration / 2)
            .attr("height", (d) => yScale(d[0]) - yScale(d[1]))
            .attr("y", (d) => yScale(d[1]));

        // @ts-ignore: TODO: Work out how to fix this
        renderCanvas(canvas, renderVirtualCanvas, width, height, update);
    }, [x, ys, data, xScale, yScale, layer, animationDuration, onMouseOver, onMouseOut, onClick]);

    return null;
}
