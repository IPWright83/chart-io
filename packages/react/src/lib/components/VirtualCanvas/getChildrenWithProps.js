import React from "react";

/**
 * Adds some additional props all the children
 * @param  {Array} options.children                     The children of the component
 * @param  {Function} options.renderVirtualCanvas       Function to call to update the virtual canvas
 * @return {Array}                                      The new children
 */
export const getChildrenWithProps = ({ children, renderVirtualCanvas }) => {
    if (children.length > 0) {
        return children.map((child) => {
            if (React.isValidElement(child)) {
                return React.cloneElement(child, { renderVirtualCanvas });
            }

            return child;
        });
    }

    if (React.isValidElement(children)) {
        return React.cloneElement(children, { renderVirtualCanvas });
    }

    return children;
};
