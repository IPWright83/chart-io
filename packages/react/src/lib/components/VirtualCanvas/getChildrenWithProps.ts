import type { IOnMouseOver, IOnMouseOut, IOnClick } from "@d3-chart/types";
import React from "react";

export type renderVirtualCanvasFunc = (
    update: d3.Transition<Element, unknown, any, unknown>,
    events: {
        onMouseOver: IOnMouseOver;
        onMouseOut: IOnMouseOut;
        onClick: IOnClick;
    }
) => void;

/**
 * Adds some additional props all the children
 * @param  children                    The children of the component
 * @param  renderVirtualCanvas         Function to call to update the virtual canvas
 * @return                             The new children
 */
export function getChildrenWithProps(children: any, renderVirtualCanvas: renderVirtualCanvasFunc): any {
    if (!children) {
        return children;
    }

    if (children.length > 0) {
        return children.map((child) => {
            if (React.isValidElement(child)) {
                // @ts-expect-error: We're dealing with really dynamic stuff here we can't really type against
                return React.cloneElement(child, { renderVirtualCanvas });
            }

            return child;
        });
    }

    if (React.isValidElement(children)) {
        // @ts-expect-error: We're dealing with really dynamic stuff here we can't really type against
        return React.cloneElement(children, { renderVirtualCanvas });
    }

    return children;
}
