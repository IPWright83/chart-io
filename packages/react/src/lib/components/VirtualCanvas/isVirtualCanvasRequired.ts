import type { ReactChildren } from "react";

/**
 * Determine if a Virtual Canvas is required or not
 * @param  children  The child or children of the virtual canvas
 * @return           True if a virtual canvas is required, otherwise false
 */
export const isVirtualCanvasRequired = (children: any): boolean => {
    if (!children) {
        return false;
    }

    // Array of children
    if (children.length) {
        return children.filter((c) => c?.type?.requiresVirtualCanvas).length > 0;
    }

    // Single child
    return children.requiresVirtualCanvas || children.type?.requiresVirtualCanvas;
};
