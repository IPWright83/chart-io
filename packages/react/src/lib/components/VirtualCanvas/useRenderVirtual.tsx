import type { IOnMouseOver, IOnMouseOut, IOnClick, IDatum } from "@d3-chart/types";
import { debounce } from "lodash";
import { useRef } from "react";
import { useStore, useSelector } from "react-redux";
import { chartSelectors, IState } from "../../store";
import {
    addEventHandlers,
    clearVirtualCanvas,
    removeEventHandlers,
    renderVirtualCanvas,
    IColorToDataMap,
} from "../../hoc/canvas/virtual";

import type { IRenderVirtualCanvasFunc } from "./getChildrenWithProps";

export const VIRTUAL_CANVAS_DEBOUNCE = 100;

export function useRenderVirtual(
    canvas?: HTMLCanvasElement,
    onMouseOver?: IOnMouseOver,
    onMouseOut?: IOnMouseOut,
    onClick?: IOnClick
) {
    // This is going to be used for the main color -> datum lookup.
    // We need to use a useRef so dependencies (wiring up events) don't re-occur forcing a re-render loop
    const colorToData = useRef<IColorToDataMap>({});

    // EventHandlers get added with a reference to colorToData, however
    // each time this changes we need to replace the event handlers. So keep
    // a reference so we can remove them
    const clickEventHandler = useRef<(e: MouseEvent) => void>();
    const moveEventHandler = useRef<(e: MouseEvent) => void>();

    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const store = useStore();

    let nodes = [];

    // Render all the virtual nodes - this is debounced to ensure that we only trigger it once
    // after all of the child layers finished their render as we don't want layers to overwrite each other
    const renderAllVirtualNodes = debounce(async () => {
        console.log("renderAllVirtualNodes", canvas);
        if (!canvas) {
            return;
        }

        // Clear and re-render the virtual canvas
        clearVirtualCanvas(canvas, width, height);
        colorToData.current = await renderVirtualCanvas(canvas, width, height, nodes);
        nodes = [];

        // Remove the old event handlers
        removeEventHandlers(canvas, clickEventHandler.current, moveEventHandler.current);

        // Replace the old canvas handlers
        const { clickHandler, moveHandler } = addEventHandlers(canvas, colorToData.current, store.dispatch);

        // Update the handler so they can be removed next time
        clickEventHandler.current = clickHandler;
        moveEventHandler.current = moveHandler;
    }, VIRTUAL_CANVAS_DEBOUNCE);

    // Whenever a child (canvas layer) renders it'll call this renderVirtual function
    // at the end of its render loop. We need to ensure that all nodes (virtual dom elements)
    // exist within the dataset and then render the virtual canvas
    const renderVirtual: IRenderVirtualCanvasFunc = (update, events) => {
        nodes = [
            ...nodes,
            {
                selection: update,
                events,
            },
        ];

        renderAllVirtualNodes();
    };

    return renderVirtual;
}
