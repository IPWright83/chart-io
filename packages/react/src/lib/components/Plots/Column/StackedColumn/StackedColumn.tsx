import { chartSelectors, IState } from "@chart-io/core";

import React from "react";
import { useSelector } from "react-redux";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { IStackedColumnBaseProps, StackedColumnBase } from "./StackedColumnBase";

export interface IStackedColumnProps extends Omit<IStackedColumnBaseProps, "interactive" | "layer"> {
    /**
     * Should Canvas be used instead of SVG?
     */
    useCanvas?: boolean;
}

const StackedCanvasColumn = withCanvas(withXYPlot<IStackedColumnProps>(StackedColumnBase), "plot stacked-column");
const StackedSVGColumn = withSVG(withXYPlot<IStackedColumnProps>(StackedColumnBase), "plot stacked-column");

/**
 * Represents a Column plot
 * @param  useCanvas   Should Canvas be used instead of SVG?
 * @param  props       The set of React properties
 * @return             The Column plot component
 */
export function StackedColumn({ useCanvas, colors, ...props }: IStackedColumnProps) {
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const palette = colors || theme.series.colors;

    if (useCanvas) {
        return <StackedCanvasColumn {...props} colors={palette} />;
    }

    return <StackedSVGColumn {...props} colors={palette} />;
}

StackedColumn.requiresVirtualCanvas = true;
StackedColumn.isPlot = true;
StackedColumn.brush = {
    horizontal: true,
    vertical: false,
};
StackedColumn.zoom = {
    horizontal: true,
    vertical: false,
};
