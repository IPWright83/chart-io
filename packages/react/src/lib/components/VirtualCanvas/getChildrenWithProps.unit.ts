import React from "react";

import { getChildrenWithProps } from "./getChildrenWithProps";

describe("VirtualCanvas getChildrenWithProps", () => {
    const renderVirtualCanvas = jest.fn();

    it("correctly augments props on a single child", () => {
        const mockElement = React.createElement("div");

        const child = getChildrenWithProps(mockElement, renderVirtualCanvas)[0];

        expect(child.type).toBe(mockElement.type);
        expect(child.props.renderVirtualCanvas).toBe(renderVirtualCanvas);
    });

    it("correctly augments a set of children", () => {
        const mockElement1 = React.createElement("div");
        const mockElement2 = React.createElement("p");

        const children = getChildrenWithProps([mockElement1, mockElement2], renderVirtualCanvas);

        expect(children.length).toBe(2);

        const child1 = children[0];
        expect(child1.type).toBe(mockElement1.type);
        expect(child1.props.renderVirtualCanvas).toBe(renderVirtualCanvas);

        const child2 = children[1];
        expect(child2.type).toBe(mockElement2.type);
        expect(child2.props.renderVirtualCanvas).toBe(renderVirtualCanvas);
    });

    it("correctly skips non React elements", () => {
        const notMockElement = {};

        const children = getChildrenWithProps([notMockElement], renderVirtualCanvas);

        expect(children.length).toBe(1);

        const child = children[0];
        expect(child).toBe(notMockElement);
        expect(child).toEqual({});
    });
});
