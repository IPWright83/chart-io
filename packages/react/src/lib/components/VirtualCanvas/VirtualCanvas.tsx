import { debounce } from "lodash";
import React, { useEffect, useCallback, useRef } from "react";
import { useStore, useSelector } from "react-redux";

import {
    addEventHandlers,
    clearVirtualCanvas,
    removeEventHandlers,
    renderVirtualCanvas,
    IColorToDataMap,
} from "../../hoc/canvas/virtual";
import { chartSelectors, IState } from "../../store";

import { getChildrenWithProps } from "./getChildrenWithProps";
import { isVirtualCanvasRequired } from "./isVirtualCanvasRequired";
import type { IRenderVirtualCanvasFunc } from "./getChildrenWithProps";

export const VIRTUAL_CANVAS_DEBOUNCE = 100;

export interface IVirtualCanvasProps {
    /**
     * The plots that are children of the virtual canvas
     */
    children: JSX.Element;
}

/**
 * The virtual canvas, draws elements to a non dom canvas and is used to
 * simulate mouse events on these elements
 * @param  {Object} props   The react props
 * @return {ReactElement}   A virtual canvas to add mouse events to canvas layers
 */
export function VirtualCanvas({ children }: IVirtualCanvasProps) {
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const store = useStore();

    // This is going to be used for the main color -> datum lookup.
    // We need to use a useRef so dependencies (wiring up events) don't re-occur forcing a re-render loop
    const colorToData = useRef<IColorToDataMap>({});
    const canvasRef = useRef<HTMLCanvasElement>();

    // We can't trigger a useEffect when the colorToData ref changes so
    // instead we use a callback function to ensure we always grab the
    // correct reference when trying to do our data lookups in the
    // addEventHandlers function
    const getColorMap = useCallback(() => colorToData.current, [colorToData.current]);

    // Adds event handlers to the canvas for triggering events
    useEffect(() => {
        const canvas = canvasRef.current;
        if (!canvas || !store) {
            return;
        }

        const { clickHandler, moveHandler } = addEventHandlers(canvas, getColorMap, store.dispatch);
        return () => {
            removeEventHandlers(canvas, clickHandler, moveHandler);
        };
    }, [canvasRef.current, store, getColorMap]);

    // This is a set of transient nodes, use as part of a debounced render function
    let nodes = [];

    // Render all the virtual nodes - this is debounced to ensure that we only trigger it once
    // after all of the child layers finished their render as we don't want layers to overwrite each other
    const renderAllVirtualNodes = debounce(async () => {
        const canvas = canvasRef.current;
        if (!canvas) {
            return;
        }

        // Clear and re-render the virtual canvas
        clearVirtualCanvas(canvas, width, height);
        colorToData.current = await renderVirtualCanvas(canvas, width, height, nodes);
        nodes = [];
    }, VIRTUAL_CANVAS_DEBOUNCE);

    // Whenever a child (canvas layer) renders it'll call this renderVirtual function
    // at the end of its render loop. We need to ensure that all nodes (virtual dom elements)
    // exist within the dataset and then render the virtual canvas
    const renderVirtual: IRenderVirtualCanvasFunc = (update, events) => {
        nodes = [
            ...nodes,
            {
                selection: update,
                events,
            },
        ];

        renderAllVirtualNodes();
    };

    // Many layers don't require the virtual canvas. If
    // they are all of these types then disable the canvas
    const includeVirtualCanvas = isVirtualCanvasRequired(children);
    if (includeVirtualCanvas === false) {
        return <React.Fragment>{children}</React.Fragment>;
    }

    // We need to extend the child components to provide the renderVirtual callback down
    // which will automatically be called by any Canvas based layers
    const childrenWithProps = getChildrenWithProps(children, renderVirtual);

    const style = {
        position: "absolute" as const,
        opacity: 0,
        zindex: 10000,
    };

    return (
        <React.Fragment>
            {childrenWithProps}
            <foreignObject className="fo" width={width} height={height}>
                <canvas className="virtual-canvas" width={width} height={height} ref={canvasRef} style={style}></canvas>
            </foreignObject>
        </React.Fragment>
    );
}
