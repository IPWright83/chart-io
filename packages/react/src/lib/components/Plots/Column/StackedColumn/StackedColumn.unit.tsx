import * as d3 from "d3";
import React from "react";
import { toMatchImageSnapshot } from "jest-image-snapshot";

import { VirtualCanvas, VIRTUAL_CANVAS_DEBOUNCE } from "../../../VirtualCanvas";
import { StackedColumn } from "./StackedColumn";

expect.extend({ toMatchImageSnapshot });

import { getBuffer, wait, renderChart, testMouseClick, testMouseOver, testMouseExit } from "../../../../testUtils";

describe("StackedColumn", () => {
    const expectedDatum = {
        x: "A",
        y: 5,
        y2: 7,
    };

    const data = [
        { x: "A", y: 5, y2: 7 },
        { x: "B", y: 10, y2: 15 },
    ];

    const scales = {
        x: d3.scaleBand().domain(["A", "B"]).range([0, 100]),
        y: d3.scaleLinear().domain([0, 30]).range([100, 0]),
        y2: d3.scaleLinear().domain([0, 30]).range([100, 0]),
    };

    describe("using SVG", () => {
        it("should render correctly", async () => {
            const { asFragment } = await renderChart({
                children: <StackedColumn x="x" ys={["y", "y2"]} />,
                data,
                scales,
            });

            expect(asFragment()).toMatchSnapshot();
        });

        describe("should handle event", () => {
            it("mouseover correctly", async () => {
                const onMouseOver = jest.fn();

                const { container, store } = await renderChart({
                    children: <StackedColumn x="x" ys={["y", "y2"]} onMouseOver={onMouseOver} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseOver(container, "rect", onMouseOver, expectedDatum);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                expect(dispatchCalls).toEqual([
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                ]);
            });

            it("mouseexit correctly", async () => {
                const onMouseOut = jest.fn();

                const { container, store } = await renderChart({
                    children: <StackedColumn x="x" ys={["y", "y2"]} onMouseOut={onMouseOut} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseExit(container, "rect", onMouseOut, expectedDatum);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                expect(dispatchCalls).toEqual([
                    // Mouseexit
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_DROPLINE",
                ]);
            });

            it("click correctly", async () => {
                const onClick = jest.fn();

                const { container, store } = await renderChart({
                    children: <StackedColumn x="x" ys={["y", "y2"]} onClick={onClick} />,
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                testMouseClick(container, "rect", onClick, expectedDatum);
            });
        });

        describe("should skip rendering if", () => {
            it("there is no x scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <StackedColumn x="x" ys={["y", "y2"]} />,
                    data,
                    scales: { y: scales.y },
                });

                expect(asFragment()).toMatchSnapshot();
            });

            it("there is no y scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <StackedColumn x="x" ys={["y", "y2"]} />,
                    data,
                    scales: { x: scales.x },
                });

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });

    describe("using Canvas", () => {
        it("should render correctly", async () => {
            const { container } = await renderChart({
                children: (
                    <VirtualCanvas>
                        <StackedColumn x="x" ys={["y", "y2"]} useCanvas={true} />
                    </VirtualCanvas>
                ),
                data,
                scales,
            });

            await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

            const canvasBuffer = getBuffer(container.querySelector(".canvas"));
            expect(canvasBuffer).toMatchImageSnapshot();

            const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
            expect(virtualCanvasBuffer).toMatchImageSnapshot();
        });

        describe("should handle event", () => {
            it("mouseover correctly", async () => {
                const onMouseOver = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas onMouseOver={onMouseOver}>
                            <StackedColumn x="x" ys={["y", "y2"]} onMouseOver={onMouseOver} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseOver(container, ".virtual-canvas", onMouseOver, expectedDatum, {
                    bubbles: true,
                    pageX: 25,
                    pageY: 95,
                });

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                expect(dispatchCalls).toEqual([
                    "EVENT.MOUSE_MOVE",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                ]);
            });

            it("mouseexit correctly", async () => {
                const onMouseOut = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas onMouseOut={onMouseOut}>
                            <StackedColumn x="x" ys={["y", "y2"]} onMouseOut={onMouseOut} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseExit(
                    container,
                    ".virtual-canvas",
                    onMouseOut,
                    expectedDatum,
                    {
                        pageX: 25,
                        pageY: 95,
                    },
                    {
                        pageX: 195,
                        pageY: 95,
                    }
                );

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);

                expect(dispatchCalls).toEqual([
                    "EVENT.MOUSE_MOVE",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.MOUSE_MOVE",
                    "EVENT.SET_TOOLTIP_COLOR",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_TOOLTIP_ITEM",
                    "EVENT.REMOVE_DROPLINE",
                ]);
            });

            it("click correctly", async () => {
                const onClick = jest.fn();

                const { container, store } = await renderChart({
                    children: (
                        <VirtualCanvas onClick={onClick}>
                            <StackedColumn x="x" ys={["y", "y2"]} onClick={onClick} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                jest.spyOn(store, "dispatch");
                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                await testMouseClick(container, ".virtual-canvas", onClick, expectedDatum, {
                    bubbles: true,
                    pageX: 25,
                    pageY: 95,
                });
            });
        });
    });
});
