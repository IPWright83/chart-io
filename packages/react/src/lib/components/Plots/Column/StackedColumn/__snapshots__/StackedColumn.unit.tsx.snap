// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackedColumn using SVG should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot stacked-column"
    >
      <g>
        <rect
          class="column"
          height="0"
          style="fill: #99c1dc; stroke: #ffffff; opacity: 0.7;"
          width="50"
          x="0"
          y="100"
        />
        <rect
          class="column"
          height="0"
          style="fill: #99c1dc; stroke: #ffffff; opacity: 0.7;"
          width="50"
          x="50"
          y="100"
        />
      </g>
      <g>
        <rect
          class="column"
          height="0"
          style="fill: #fc998e; stroke: #ffffff; opacity: 0.7;"
          width="50"
          x="0"
          y="100"
        />
        <rect
          class="column"
          height="0"
          style="fill: #fc998e; stroke: #ffffff; opacity: 0.7;"
          width="50"
          x="50"
          y="100"
        />
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`StackedColumn using SVG should skip rendering if there is no x scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot stacked-column"
    />
  </svg>
</DocumentFragment>
`;

exports[`StackedColumn using SVG should skip rendering if there is no y scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot stacked-column"
    />
  </svg>
</DocumentFragment>
`;
